packages/vite/src/node/ssr/runtime/__tests__/utils.ts

```diff
@@ -0,0 +1,122 @@
+import fs from 'node:fs'
+import { dirname, resolve } from 'node:path'
+import { fileURLToPath } from 'node:url'
+import type { TestAPI } from 'vitest'
+import { afterEach, beforeEach, test } from 'vitest'
+import type { InlineConfig, ViteDevServer } from '../../../index'
+import { createServer } from '../../../index'
+import type { ViteRuntime } from '../runtime'
+import { createViteRuntime } from '../node/mainThreadRuntime'
+
+interface TestClient {
+  server: ViteDevServer
+  runtime: ViteRuntime
+}
+
+export async function createViteRuntimeTester(
+  config: InlineConfig = {},
+): Promise<TestAPI<TestClient>> {
+  function waitForWatcher(server: ViteDevServer) {
+    return new Promise<void>((resolve) => {
+      if ((server.watcher as any)._readyEmitted) {
+        resolve()
+      } else {
+        server.watcher.once('ready', () => resolve())
+      }
+    })
+  }
+
+  beforeEach<TestClient>(async (t) => {
+    globalThis.__HMR__ = {}
+
+    t.server = await createServer({
+      root: __dirname,
+      logLevel: 'error',
+      server: {
+        middlewareMode: true,
+        watch: null,
+        hmr: {
+          port: 9609,
+        },
+      },
+      ssr: {
+        external: ['@vitejs/cjs-external', '@vitejs/esm-external'],
+      },
+      optimizeDeps: {
+        disabled: true,
+        noDiscovery: true,
+        include: [],
+      },
+      plugins: [
+        {
+          name: 'vite-plugin-virtual',
+          resolveId(id) {
+            if (id === 'virtual0:test') {
+              return `\0virtual:test`
+            }
+            if (id === 'virtual:test') {
+              return 'virtual:test'
+            }
+          },
+          load(id) {
+            if (id === `\0virtual:test`) {
+              return `export const msg = 'virtual0'`
+            }
+            if (id === `virtual:test`) {
+              return `export const msg = 'virtual'`
+            }
+          },
+        },
+      ],
+      ...config,
+    })
+    t.runtime = await createViteRuntime(t.server, {
+      hmr: {
+        logger: false,
+      },
+      sourcemapInterceptor: 'prepareStackTrace',
```

Comment:
I noticed that this is breaking Vitest's own stacktrace fixing system, but I suppose this is expected since both makes use of `Error.prepareStackTrace`. I just thought it's worth mentioning since I was a little confused when debugging something in the unit tests, for example:

<details><summary>Reveal example</summary>

```ts
  it.only("xxx", async ({ runtime }) => {
    await runtime.executeEntrypoint("./not-found"); // line 268
  })
```

will show:

```
 FAIL  packages/vite/src/node/ssr/runtime/__tests__/server-runtime.spec.ts > vite-runtime initialization > xxx
Error: Failed to load url ./not-found (resolved id: ./not-found). Does the file exist?
 ❯ loadAndTransform packages/vite/src/node/server/transformRequest.ts:161:17
    159|   }
    160| 
    161|   const resolved = module
       |                 ^
    162|     ? undefined
    163|     : (await pluginContainer.resolveId(url, undefined, { ssr })) ?? undefined
 ❯ Module.fetchModule packages/vite/src/node/ssr/fetchModule.ts:53:16
 ❯ ViteRuntime.cachedModule packages/vite/src/node/ssr/runtime/runtime.ts:158:91
 ❯ ViteRuntime.executeEntrypoint packages/vite/src/node/ssr/runtime/runtime.ts:64:27
 ❯ packages/vite/src/node/ssr/runtime/__tests__/server-runtime.spec.ts:232:5
```

while `sourcemapInterceptor: false` would show:

```
 FAIL  packages/vite/src/node/ssr/runtime/__tests__/server-runtime.spec.ts > vite-runtime initialization > xxx
Error: Failed to load url ./not-found (resolved id: ./not-found). Does the file exist?
 ❯ loadAndTransform packages/vite/src/node/server/transformRequest.ts:272:22
    270|       .next().value
    271|     const importer = importerMod?.file || importerMod?.url
    272|     const err: any = new Error(
       |                      ^
    273|       `Failed to load url ${url} (resolved id: ${id})${
    274|         importer ? ` in ${importer}` : ''
 ❯ Module.fetchModule packages/vite/src/node/ssr/fetchModule.ts:80:16
 ❯ ViteRuntime.cachedModule packages/vite/src/node/ssr/runtime/runtime.ts:239:9
 ❯ ViteRuntime.executeEntrypoint packages/vite/src/node/ssr/runtime/runtime.ts:98:27
 ❯ packages/vite/src/node/ssr/runtime/__tests__/server-runtime.spec.ts:268:5
```

</details>

Reply:
Yeah, I'm not sure if we can support both somehow. Every new runtime will override it.

Maybe we can have a shared cache somehow so when vite-node switches to this implementation, it could still work

Reply:
I would love to see if there is a clever way to make it work, but I suppose this is only a minor inconvenience for vite developers, so probably it doesn't much putting a huge effort on this.

One thing I was thinking is that, maybe you can move stacktrace related tests into a dedicated spec file like `server-runtime-stacktrace.spec.ts`, then for other tests, `sourcemapInterceptor: false` can be used to benefit from Vitest stacktrace.

Reply:
Interceptor now stores runtime's module caches (there is now `destroy` method to clear that side effect). I also refactored source map tests into a separate file 👍🏻